Add mob conversion event

Syst3ms opened the pull request:
While working on #3476, I was surprised to find there is no event for mob conversion. This is a generally useful mechanic to expose, so this PR adds a notification event to listen for it. There is no "allow" event as mob conversion happens in a variety of contexts, and what it means to "cancel" conversion can mean a lot of very different things.

Syst3ms committed:
Add mob conversion event

Syst3ms committed:
spacing

LlamaLad7 commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/mixin/entity/event/MobEntityMixin.java:
```diff
@@ -0,0 +1,36 @@
+@Mixin(MobEntity.class)
+public class MobEntityMixin {
+	@Inject(method = "convertTo", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/World;spawnEntity(Lnet/minecraft/entity/Entity;)Z"))
+	private <T extends MobEntity> void afterEntityConverted(EntityType<T> entityType, boolean keepEquipment, CallbackInfoReturnable<T> cir, @Local T converted) {
```
Prefer `@ModifyArg` over local capture. You can get `keepEquipment` with `@Local(argsOnly = true)`

Syst3ms replied:
Ohhh, that's quite sneaky, hadn't thought of that.

Syst3ms committed:
use a smarter mixin strategy

Syst3ms committed:
add pig and villager lightning triggers

haykam821 commented on fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerLivingEntityEvents.java:
```diff
@@ -73,6 +74,18 @@ public final class ServerLivingEntityEvents {
+		for (MobConversion callback : callbacks) {
+			callback.onConversion(previous, converted, keepEquiment);
+		}
+	});
```
The `keepEquiment` parameter name contains a typo:
 
```suggestion
	public static final Event<MobConversion> MOB_CONVERSION = EventFactory.createArrayBacked(MobConversion.class, callbacks -> (previous, converted, keepEquipment) -> {
		for (MobConversion callback : callbacks) {
			callback.onConversion(previous, converted, keepEquipment);
		}
	});
```

Syst3ms committed:
tyop

Co-authored-by: haykam821 <24855774+haykam821@users.noreply.github.com>

modmuss50 commented:
Is the plan to get this in before the data attachment API? I think it makes sense to?

Syst3ms commented:
Yes, preferrably, that avoids having to make a secondary PR to fix this in data attachment.

modmuss50 commented:
Added to last call, should only need a few days at most 👍

Shnupbups commented:
A special case from vanilla that isn't caught by this is Tadpoles growing up into Frogs, in the `TadpoleEntity#growUp()` method. They are indeed separate entity types, and it utilizes `SpawnReason.CONVERSION` so I believe this is justified to fit this event.

Shnupbups commented:
Another issue raised on Discord is that perhaps there needs to be clear JavaDoc explaining that this event only pertains to when the resultant entity is different from the original mob, perhaps using the Red Mooshroom to Brown Mooshroom transformation as an example of something that is explicitly NOT covered.

maityyy commented:
Charged Creeper is also an example

Syst3ms committed:
handle tadpole -> frog conversion

- update javadoc to clarify the handled cases
- document keepEquipment parameter
- improve mixin handler names
